fem/lor.hpp

```diff
@@ -0,0 +1,287 @@
+public:
+   /// @brief Create a solver of type @a SolverType, formed using the assembled
+   /// SparseMatrix of the %LOR version of @a a_ho. @see LOR
+   LORSolver(BilinearForm &a_ho, const Array<int> &ess_tdof_list,
+             int ref_type=BasisType::GaussLobatto)
+   {
+      lor = new LOR(a_ho, ess_tdof_list, ref_type);
+      SetOperator(*lor->GetAssembledSystem());
+   }
+
+#ifdef MFEM_USE_MPI
+   /// @brief Create a solver of type @a SolverType, formed using the assembled
+   /// HypreParMatrix of the %LOR version of @a a_ho. @see ParLOR
+   LORSolver(ParBilinearForm &a_ho, const Array<int> &ess_tdof_list,
+             int ref_type=BasisType::GaussLobatto)
+   {
+      lor = new ParLOR(a_ho, ess_tdof_list, ref_type);
+      SetOperator(*lor->GetAssembledSystem());
+   }
+#endif
+
+   /// @brief Create a solver of type @a SolverType using Operator @a op and
+   /// arguments @a args.
+   ///
+   /// The object @a lor_ will be used for DOF permutations.
+   template <typename... Args>
+   LORSolver(const Operator &op, LORBase &lor_, Args&&... args) : solver(args...)
+   {
+      lor = &lor_;
+      own_lor = false;
+      SetOperator(op);
+   }
+
+   /// @brief Create a solver of type @a SolverType using the assembled %LOR
+   /// operator represented by @a lor_.
+   ///
+   /// The given @a args will be used as arguments to the solver constructor.
+   template <typename... Args>
+   LORSolver(LORBase &lor_, Args&&... args)
+      : LORSolver(*lor_.GetAssembledSystem(), lor_, args...) { }
+
+   void SetOperator(const Operator &op)
+   {
+      solver.SetOperator(op);
+      width = solver.Width();
+      height = solver.Height();
+   }
+
+   void Mult(const Vector &x, Vector &y) const
+   {
+      if (use_permutation && lor->RequiresDofPermutation())
+      {
+         const Array<int> &p = lor->GetDofPermutation();
+         px.SetSize(x.Size());
+         py.SetSize(y.Size());
+         for (int i=0; i<x.Size(); ++i)
+         { px[i] = p[i] < 0 ? -x[-1-p[i]] : x[p[i]]; }
+
+         solver.Mult(px, py);
+
+         for (int i=0; i<y.Size(); ++i)
+         {
+            int pi = p[i];
+            int s = pi < 0 ? -1 : 1;
+            y[pi < 0 ? -1-pi : pi] = s*py[i];
+         }
+      }
+      else
+      {
+         solver.Mult(x, y);
+      }
+   }
+
+   /// @brief Enable or disable the DOF permutation (enabled by default).
+   ///
+   /// The corresponding %LOR and high-order DOFs may not have the same
+   /// numbering (for example, when using ND or RT spaces), and so a permutation
+   /// is required when applying the %LOR solver as a preconditioner for the
+   /// high-order problem. This permutation can be disabled (for example, in
+   /// order to precondition the low-order problem directly).
+   void UsePermutation(bool use_permutation_)
+   {
+      use_permutation = use_permutation_;
+   }
+
+   /// Support the use of -> to call methods of the underlying solver.
+   SolverType *operator->() const { return solver; }
```

Comment:
If I understand this correctly, for a `LORSolver *s` pointer, `s->Mult` would call `LORSolver::Mult`; for a `LORSolver s` object, `s->Mult` would call `s.solver.Mult`. If that is correct, it seems perhaps convenient but also potentially confusing and likely to cause bugs (which may be hard to track down, i.e. it may run but not apply the permutation). Would it be clearer and safer to just have a `GetSolver()` function?

Reply:
The purpose of overloading the `->` operator is to allow for setting solver-specific parameters, e.g.
```
LORSolver<HypreBoomerAMG> lor_solver(…);
lor_solver->SetRelaxType(…);
```

Of course, if you call `lor_solver->Mult` in this case, it will do the wrong thing, which may be a risk for confusion.

Reply:
So if we replace this function with `GetSolver`, then `lor_solver->SetRelaxType(…);` becomes `lor_solver.GetSolver()->SetRelaxType(…);`. I think that would be safer.

Reply:
Ok, I removed the `->` operator, and added `GetSolver` (which returns a reference, rather than a pointer, since the underlying solver can never be NULL), so that now it will be `lor_solver.GetSolver().SetRelaxType(…);`